pyadjoint/optimization/tao_solver.py

```diff
@@ -0,0 +1,491 @@
+from functools import cached_property
+from numbers import Complex
+import weakref
+
+import numpy as np
+
+from ..enlisting import Enlist
+from ..overloaded_type import OverloadedType
+from .optimization_problem import MinimizationProblem
+from .optimization_solver import OptimizationSolver
+
+
+try:
+    import petsc4py.PETSc as PETSc
+except ModuleNotFoundError:
+    PETSc = None
+
+__all__ = \
+    [
+        "TAOSolver"
+    ]
+
+
+def attach_destroy_finalizer(obj, *args):
+    def finalize_callback(*args):
+        for arg in args:
+            if arg is not None:
+                arg.destroy()
+
+    finalize = weakref.finalize(obj, finalize_callback,
+                                *args)
+    finalize.atexit = False
+
+
+class PETScVecInterface:
+    """Interface for conversion between :class:`OverloadedType` objects and
+    :class:`petsc4py.PETSc.Vec` objects.
+
+    This uses the generic interface provided by :class:`OverloadedType`.
+
+    Args:
+        X (OverloadedType or Sequence[OverloadedType]): Defines the data
+            layout.
+        comm (petsc4py.PETSc.Comm or mpi4py.MPI.Comm): Communicator.
+    """
+
+    def __init__(self, X, *, comm=None):
+        if PETSc is None:
+            raise RuntimeError("PETSc not available")
+
+        X = Enlist(X)
+        if comm is None:
+            comm = PETSc.COMM_WORLD
+        if hasattr(comm, "tompi4py"):
+            comm = comm.tompi4py()
+
+        vecs = tuple(x._ad_to_petsc() for x in X)
+        n = sum(vec.getLocalSize() for vec in vecs)
+        N = sum(vec.getSize() for vec in vecs)
+        _, isets = PETSc.Vec().concatenate(vecs)
+        for vec in vecs:
```

Comment:
If the destructor is really not necessary, let's not do it.

if it is necessary then `_ad_to_petsc()` should probably be a context manager so as to remove the footgun.

Reply:
Removed.